core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -309,8 +309,8 @@ public CreatePathResult createPath(AlluxioURI path, CreatePathOptions options)
         // to the non-persisted Inodes of traversalResult.
         traversalResult.getNonPersisted().add(lastInode);
         toPersistDirectories.add(lastInode);
-      } else if (!(options instanceof CreateDirectoryOptions && ((CreateDirectoryOptions) options)
-          .isAllowExists())) {
+      } else if (!(lastInode.isDirectory() && options instanceof CreateDirectoryOptions
```

Comment:
I think this would be easier to reason about with the `!` distributed since then it's clear that each condition on it's own should cause `FileAlreadyExistsException` to be thrown.